sdk/identity/azure-identity/src/azure_pipelines_credential.cpp

```diff
@@ -0,0 +1,257 @@
+
+  bool isTenantIdValid = IsValidTenantId(tenantId);
+  if (!isTenantIdValid)
+  {
+    IdentityLog::Write(
+        IdentityLog::Level::Warning,
+        "Invalid tenant ID provided  for " + GetCredentialName()
+            + ". The tenant ID must be a non-empty string containing only alphanumeric characters, "
+              "periods, or hyphens. You can locate your tenantID by following the instructions "
+              "listed here: https://learn.microsoft.com/partner-center/find-ids-and-domain-names");
+  }
+  if (clientId.empty())
+  {
+    IdentityLog::Write(
+        IdentityLog::Level::Warning, "No client ID specified for " + GetCredentialName() + ".");
+  }
+  if (serviceConnectionId.empty())
+  {
+    IdentityLog::Write(
+        IdentityLog::Level::Warning,
+        "No service connection ID specified for " + GetCredentialName() + ".");
+  }
+  if (systemAccessToken.empty())
+  {
+    IdentityLog::Write(
+        IdentityLog::Level::Warning,
+        "No system access token specified for " + GetCredentialName() + ".");
+  }
+  if (m_oidcRequestUrl.empty())
+  {
+    IdentityLog::Write(
+        IdentityLog::Level::Warning,
+        "No value for environment variable '" + Azure::Identity::_detail::OidcRequestUrlEnvVarName
+            + "' needed by " + GetCredentialName() + ". This should be set by Azure Pipelines.");
+  }
+
+  if (isTenantIdValid && !clientId.empty() && !serviceConnectionId.empty()
+      && !systemAccessToken.empty() && !m_oidcRequestUrl.empty())
+  {
+    m_tokenCredentialImpl = std::make_unique<TokenCredentialImpl>(options);
+    m_requestBody
+        = std::string(
+              "grant_type=client_credentials"
+              "&client_assertion_type="
+              "urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer" // cspell:disable-line
+              "&client_id=")
+        + Url::Encode(clientId);
+
+    IdentityLog::Write(
+        IdentityLog::Level::Informational, GetCredentialName() + " was created successfully.");
+  }
+  else
+  {
+    // Rather than throwing an exception in the ctor, following the pattern in existing credentials
+    // to log the errors, and defer throwing an exception to the first call of GetToken(). This is
+    // primarily needed for credentials that are part of the DefaultAzureCredential, which this
+    // credential is not intended for.
+    IdentityLog::Write(
+        IdentityLog::Level::Warning,
+        "Azure Pipelines environment is not set up for the " + GetCredentialName()
+            + " credential to work.");
+  }
+}
+
+Request AzurePipelinesCredential::CreateOidcRequestMessage() const
+{
+  const std::string oidcApiVersion = "7.1";
+
+  Url requestUrl = Url(
+      m_oidcRequestUrl + "?api-version=" + Url::Encode(oidcApiVersion)
+      + "&serviceConnectionId=" + Url::Encode(m_serviceConnectionId));
+  Request request = Request(HttpMethod::Post, requestUrl);
+  request.SetHeader("content-type", "application/json");
```

Comment:
Do you need to set this? The request has no body

Reply:
I saw it being set in .NET. Maybe we'd want to remove it from both?
https://github.com/Azure/azure-sdk-for-net/blob/31fd5abeb571025ab5be7d7e8114e7df4a355325/sdk/identity/Azure.Identity/src/Credentials/AzurePipelinesCredential.cs#L106

cc @christothes

Reply:
I was following off of the powershell command that made the REST api token call. And they had the content-type set. Let me find it.